### PR TITLE
Write to one npy

### PR DESCRIPTION
**Pull request by dmitriy-serdyuk**
Concerns #22

@ruotianluo what do you think?

### CODE REVIEW

**Comment by ruotianluo** [comment]
Why can't I incrementally save the array?

File: scripts/prepro_feats.py [hunk 1]
```diff
@@ -50,6 +46,13 @@
 from misc.resnet_utils import myResnet
 import misc.resnet as resnet
 
+
+def append_array(f, f_toc, name, array):
+  np.save(f, array)
+  offset = f.tell()
+  f_toc.write("{} {}\n".format(name, offset))
+
```
Comment (ruotianluo) [inline scripts/prepro_feats.py]: Shouldn't you set the offset before np.save?

Comment (dmitriy-serdyuk) [reply scripts/prepro_feats.py]: Good call

**Comment by dmitriy-serdyuk** [comment]
> Why can't I incrementally save the array?
Sorry? If you mean npz, numpy doesn't provide this option. Saving with zlib is also not possible because it doesn't work with files >2G.
> How the speed of training change?
I am benchmarking it now.

File: scripts/prepro_feats.py [hunk 2]
```diff
@@ -50,6 +46,13 @@
 from misc.resnet_utils import myResnet
 import misc.resnet as resnet
 
+
+def append_array(f, f_toc, name, array):
+  offset = f.tell()
+  np.save(f, array)
+  f_toc.write("{} {}\n".format(name, offset))
```
Comment (ruotianluo) [inline scripts/prepro_feats.py]: Forget about my deleted question. You can actually do `np.savez_compressed(f, feat=array)`.

File: dataloader.py [hunk 3]
```diff
@@ -89,13 +97,27 @@ def __init__(self, opt):
...
             for split in self.iterators.keys():
                 del self._prefetch_process[split]
+
         import atexit
         atexit.register(cleanup)
 
+        toc_filename = os.path.join(self.input_fc_dir, 'dataset_fc.toc')
+        with open(toc_filename, 'rt') as f:
+            self.fc_toc = {l.split()[0]: int(l.split()[1])
+                           for l in f.readlines()}
```
Comment (ruotianluo) [inline dataloader.py]: Is it simpler just saving the offsets in a pickle file or a json file?

Comment (dmitriy-serdyuk) [reply dataloader.py]: e3b7c415b3ebded1356ab7de1550332b8a85bfa1

File: scripts/prepro_feats.py (hunk 2 above)
Comment (ruotianluo) [reply scripts/prepro_feats.py]: I do find np.savez_compressed will cause some weird bug, but I don't know why that happens.

Comment (ruotianluo) [reply scripts/prepro_feats.py]: I think I found the solution.

**Comment by ruotianluo** [comment]
I'm merging it now, and will change to compression saving.

**Comment by dmitriy-serdyuk** [comment]
Some benchmarks (batch size 32)

|                     | Feature extraction  | Per batch (local FS)| Per batch (NFS) |
|------------ | -------------|-----------|------------|
|File per feature | 3.65h | 0.3s| 0.55s|
|One big npy       | 1.88h | 0.5s| 0.10s|

There is a definite speed up in dataset creation. And it's much easier to share features. But the performance numbers are very confusing so far.

**Comment by ruotianluo** [comment]
Thanks. The 0.10s sounds to me like the data has been loaded into the memory.

**Comment by ruotianluo** [comment]
And, I would think it should be better for (one bignpy/ local fs) if you use my latest commit which compress the features.

**Comment by dmitriy-serdyuk** [comment]
I think what happens is that the file server has enough memory to cache the whole dataset (or a big chunk of it).

**Pull request merged**